src/tracks/ui/TextEditHelper.cpp

```diff
@@ -0,0 +1,666 @@
+#endif
+                }
+                else
+                    initial = position;
+
+                mInitialCursorPos = initial;
+                mCurrentCursorPos = position;
+            }
+            else
+            {
+                if (mInitialCursorPos == mCurrentCursorPos)
+                {
+                    auto position = FindCursorIndex(event.GetPosition());
+                    mInitialCursorPos = mCurrentCursorPos = position;
+                }
+                // Actually this might be right or middle down
+                mRightDragging = true;
+            }
+
+#if defined(__WXGTK__) && (HAVE_GTK)
+            if (evt.MiddleDown()) {
+                // Paste text, making a NEW label if none is selected.
+                wxTheClipboard->UsePrimarySelection(true);
+                view.PasteSelectedText(project, newSel.t0(), newSel.t1());
+                wxTheClipboard->UsePrimarySelection(false);
+}
+#endif
+            return true;
+        }
+    }
+    return false;
+}
+
+bool TextEditHelper::OnDrag(const wxMouseEvent& event, AudacityProject* project)
+{
+    return HandleDragRelease(event, project);
+}
+
+bool TextEditHelper::OnRelease(const wxMouseEvent& event, AudacityProject* project)
+{
+    return HandleDragRelease(event, project);
+}
+
+void TextEditHelper::Draw(wxDC& dc, const wxRect& rect)
+{
+    mBBox = rect;
+    dc.SetFont(mFont);
+
+    const auto cursorHeight = dc.GetFontMetrics().height;
+
+    wxRect prevClippingRegion;
+    auto restoreClippingRegion = dc.GetClippingBox(prevClippingRegion);
+    dc.SetClippingRegion(rect);
+
+    
+    auto rtl = wxTheApp->GetLayoutDirection() == wxLayout_RightToLeft;
+
+    auto curPosX = 0;
+    auto maxOffset = static_cast<int>(mText.Length());
+    mOffset = std::clamp(mOffset, 0, maxOffset);
+    {
+        auto leftBound = rect.GetLeft();
+        auto rightBound = rect.GetRight() + 1;
+        GetCharPositionX(mCurrentCursorPos, &curPosX);
+
+        if ((!rtl && curPosX >= rightBound) || (rtl && curPosX < leftBound))
+        {
+            while (mOffset < maxOffset)
+            {
+                GetCharPositionX(mCurrentCursorPos, &curPosX);
+                if (curPosX < rightBound && curPosX >= leftBound)
+                    break;
+                ++mOffset;
+            }
+        }
+        if ((!rtl && curPosX < leftBound) || (rtl && curPosX >= rightBound))
+        {
+            while (mOffset > 0)
+            {
+                GetCharPositionX(mCurrentCursorPos, &curPosX);
+                if (curPosX >= leftBound && curPosX < rightBound)
+                    break;
+                --mOffset;
+            }
+        }
+    }
+
+    if (mCurrentCursorPos != mInitialCursorPos)
+    {
+        auto left = 0;
+        auto right = 0;
+        GetCharPositionX(std::min(mCurrentCursorPos, mInitialCursorPos), &left);
+        GetCharPositionX(std::max(mCurrentCursorPos, mInitialCursorPos), &right);
```

Comment:
Are the above two lines right?  Isn't current >= initial always?

Is what you really want to do:  get char positions for initial and current, as left and right, but then swap in case of RTL layout?  Because, swap is needed in that case and that case only?

Did you really try all of this out with Arabic?  Putting breakpoints at every test of layout, and making sure all are hit?

Reply:
>Are the above two lines right? Isn't current >= initial always?

Not always, depends on direction of cursor movement

Reply:
But I thought mCurrentCursorPos and mInitialCursorPos are logical positions in a string of characters, not x coordinates on a screen.  If they are not that,  what are they?

Reply:
Yes, they are a positions in a string. mInitialCursorPos is not moving and initialized when interaction is started, then, when user moves cursor either with mouse (holding a left button), or with Shift+Arrow Left/Right mCurrentCursorPos is used to track the "current" cusror position